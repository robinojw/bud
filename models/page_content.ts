export type UserData = {
  balance: Amount;
  id: string;
  provider: Provider;
  transactions: Array<Transaction>;
}

export type Provider = {
  account_number: string;
  description: string;
  sort_code: string;
  title: string;
}

export type Amount = {
  currency: string;
  value: number;
}

export type Transaction = {
  amount: Amount;
  category_title: string;
  date: Date;
  description: string;
  id: string;
}
