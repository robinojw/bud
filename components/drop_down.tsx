import React, {FunctionComponent} from 'react';
import styled from 'styled-components';

import {Transaction} from '../models/page_content';
import {foregroundColour, colourBlue} from '../styles/global';

type DropDownProps = {
  transactions: Array<Transaction>;
  setList: (list: Array<Transaction>) => void;
};

const Select = styled.select`
  background: ${foregroundColour};
  border-radius: 4px;
  height: 30px;
  width: 170px;
  color: ${colourBlue};
`;

export const DropDown: FunctionComponent<DropDownProps> = ({
  transactions,
  setList,
}) => {
  function handleChange(event) {
    if (!event.target.value) return;
    switch (event.target.value) {
      case -1:
        setList(sortSmallest(transactions));
        break;
      case 1:
        setList(sortLargest(transactions));
        break;
      case 0:
        setList(sortRecent(transactions));
        break;
      default:
        return;
        break;
    }
  }

  return (
    <div>
      <Select onChange={handleChange}>
        <option value={-1}>Smallest Transactions</option>
        <option value={1}>Largest Transactions</option>
        <option value={0}>Most Recent</option>
      </Select>
    </div>
  );
};

export const sortSmallest = (transactions: Array<Transaction>) => {
  return transactions.sort((a, b) => {
    if (Math.abs(a.amount.value) < Math.abs(b.amount.value)) {
      return -1;
    }
    if (Math.abs(a.amount.value) > Math.abs(b.amount.value)) {
      return 1;
    }
    return 0;
  });
};

export const sortLargest = (transactions: Array<Transaction>) => {
  return transactions.sort((a, b) => {
    if (Math.abs(a.amount.value) > Math.abs(b.amount.value)) {
      return -1;
    }
    if (Math.abs(a.amount.value) < Math.abs(b.amount.value)) {
      return 1;
    }
    return 0;
  });
};

export const sortRecent = (transactions: Array<Transaction>) => {
  return transactions.sort(
    (a, b) => new Date(b.date).getTime() - new Date(a.date).getTime()
  );
};
