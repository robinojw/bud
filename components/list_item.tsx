import React, {FunctionComponent} from 'react';
import styled from 'styled-components';

import {Transaction} from '../models/page_content';
import {
  textSm,
  foregroundColour,
  colourGray,
  colourGreen,
  colourRed,
} from '../styles/global';

type ItemProps = {
  transaction: Transaction;
};

const Container = styled.div`
  background: ${foregroundColour};
  border-radius: 4px;
  display: flex;
  height: 75px;
  justify-content: space-between;
  margin-bottom: 25px;
  width: 100%;
`;

const Info = styled.div`
  display: flex;
  flex-direction: column;
  justify-content: space-between;
  padding: 15px 20px;

  p {
    margin: 0;
    padding: 0;
    text-align: ${props => (props.val ? 'right' : 'left')};
  }

  .subtext {
    color: ${colourGray};
    font: ${textSm};
    margin: 0;
    padding: 0;
  }

  .value {
    color: ${props => (props.val > 0 ? colourGreen : colourRed)};
  }
`;

export const Item: FunctionComponent<ItemProps> = ({transaction}) => {
  return (
    <Container>
      <Info>
        <p>{transaction.description}</p>
        <p className="subtext">{transaction.category_title}</p>
      </Info>
      <Info val={transaction.amount.value}>
        <p className="subtext">{transaction.date}</p>
        <p className="value">{transaction.amount.value}</p>
      </Info>
    </Container>
  );
};
