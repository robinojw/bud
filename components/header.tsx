import React, {FunctionComponent} from 'react';
import styled from 'styled-components';

import {Transaction} from '../models/page_content';
import {colourGray, textBold, textSm, colourWhite} from '../styles/global';
import {DropDown} from './drop_down';

const Container = styled.div`
  display: flex;
  justify-content: space-between;
  width: 100%;
`;

const User = styled.a`
  display: flex;
  text-decoration: none;
  color: ${colourWhite};
`;

const Name = styled.div`
  display: flex;
  flex-direction: column;
  justify-content: center;

  h2 {
    font: ${textBold};
    margin: 0 0 5px 0;
  }

  p {
    font: ${textSm};
    color: ${colourGray};
    margin: 0;
  }
`;

const ProfileImg = styled.div`
  height: 75px;
  width: 75px;
  margin-inline-end: 15px;

  img {
    height: 100%;
    object-fit: cover;
    width: 100%;
    border-radius: 50%;
  }
`;

type HeaderProps = {
  image: string;
  transactions: Array<Transaction>;
  setList: (list: Array<Transaction>) => void;
};

export const Header: FunctionComponent<HeaderProps> = ({
  image,
  transactions,
  setList,
}) => {
  return (
    <Container>
      <User
        href="https://robinw.co.uk"
        target="_blank"
        rel="noopener noreferrer"
      >
        <ProfileImg>
          <picture>
            <source srcSet={image + '.webp'} type="image/webp" />
            <img loading="lazy" alt="Profile Picture" srcSet={image + '.jpg'} />
          </picture>
        </ProfileImg>
        <Name>
          <h2>Robin White</h2>
          <p>Software Engineer</p>
        </Name>
      </User>
      <DropDown transactions={transactions} setList={setList} />
    </Container>
  );
};
