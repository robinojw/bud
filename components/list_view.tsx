import React, {FunctionComponent, useState, useEffect} from 'react';
import styled from 'styled-components';
import {Transaction} from '../models/page_content';
import {Item} from './list_item';
import {titleBold} from '../styles/global';

const listMax = 10;

type ListProps = {
  title: string;
  transactions: Array<Transaction>;
};

const List = styled.div`
  display: flex;
  flex-direction: column;
  width: 100%;
  overflow-y: scroll;
`;

const Title = styled.h2`
  font: ${titleBold};
  margin: 75px 0 25px;
`;

export const ListView: FunctionComponent<ListProps> = ({
  title,
  transactions,
}) => {
  return (
    <>
      <Title>{title}</Title>
      <List>
        {transactions.slice(0, listMax).map(index => (
          <Item key={index.id} transaction={index} />
        ))}
      </List>
    </>
  );
};
