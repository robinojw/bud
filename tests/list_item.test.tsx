import React from 'react';
import renderer from 'react-test-renderer';
import {Item} from "../components/list_item";
import {Transaction} from "../models/page_content";
import 'jest-styled-components';
import moment from 'moment';

export const mockTransaction: Transaction = {
  amount: {
    value: 10,
    currency: 'GBP'
  },
  category_title: 'Travel',
  description: 'Trainline',
  id: 'jdsalfjkejojwefd23254324',
  date: moment("2017-09-15 09:30:00").format("MMM Do YYYY h:mm:ss a"),
}

it('List View component matches snapshot', () => {
  const component = renderer.create(
    <Item transaction={mockTransaction} />
  ).toJSON();
  expect(component).toMatchSnapshot();
});
