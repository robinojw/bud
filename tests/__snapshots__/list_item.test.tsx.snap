// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`List View component matches snapshot 1`] = `
.c0 {
  background: #0A152B;
  border-radius: 4px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  height: 75px;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  margin-bottom: 25px;
  width: 100%;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  padding: 15px 20px;
}

.c1 p {
  margin: 0;
  padding: 0;
  text-align: left;
}

.c1 .subtext {
  color: #888888;
  font: 400 12px 'Roboto',sans-serif;
  margin: 0;
  padding: 0;
}

.c1 .value {
  color: #FF5252;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  padding: 15px 20px;
}

.c2 p {
  margin: 0;
  padding: 0;
  text-align: right;
}

.c2 .subtext {
  color: #888888;
  font: 400 12px 'Roboto',sans-serif;
  margin: 0;
  padding: 0;
}

.c2 .value {
  color: #8EDD5D;
}

<div
  className="c0"
>
  <div
    className="c1"
  >
    <p>
      Trainline
    </p>
    <p
      className="subtext"
    >
      Travel
    </p>
  </div>
  <div
    className="c2"
  >
    <p
      className="subtext"
    >
      Sep 15th 2017 9:30:00 am
    </p>
    <p
      className="value"
    >
      10
    </p>
  </div>
</div>
`;
